### DISCUSSION TITLE
Example with Electron

### DISCUSSION THREAD

**Discussion by MBoui** [nuxt/nuxt#16082] [Questions]
Hello the community,

I am looking for an example (starter) to using Electron with nuxt3.
Someone can point me to the good direction?

Thanks!

**Comment by nwittwer** [comment]
I'm also curious about this 🙂

There's a great Electron + Nuxt 2 project here: https://github.com/michalzaq12/electron-nuxt but it [doesn't support Nuxt 3](https://github.com/michalzaq12/electron-nuxt/issues/863) currently.

There's a few Vue 3 + Vite projects

- [electron-vite-vue](https://github.com/electron-vite/electron-vite-vue)
- [electron-vite-boilerplate](https://github.com/electron-vite/electron-vite-boilerplate)
- [electron-vue-template](https://github.com/Deluze/electron-vue-template)

... but it isn't clear how this would work with Nuxt 3.

**Reply by wobsoriano** [reply]
They won't work with Nuxt, but I guess you can fork the first [one](https://github.com/electron-vite/electron-vite-vue) and make a Nuxt module out of it

**Reply by caoxiemeihao** [reply]
I have created a Nuxt module for Electron 👉 [nuxt-electron](https://github.com/caoxiemeihao/nuxt-electron)

**Reply by ChristianPachecoLopez** [reply]
@caoxiemeihao does it also work for nuxt 4?

**Comment by CharlesBT** [comment]
+1 any advice on how to run Nuxt3 behind electron ?

**Comment by zavodnyrichard** [comment]
I'm also interested in running Nuxt3 with Electron.js

**Comment by anton-liam** [comment]
any news?

**Comment by wobsoriano** [comment]
Here's a working one in dev https://github.com/wobsoriano/nuxt-electron-example

Have to figure out how to run `.outpu`.

**Comment by gurvancampion** [comment]
Here is an example I made also only working in development for the moment: https://github.com/gurvan-guss/nuxt-electron-trpc-prisma
I'm trying to figure out how to make it work when I pack it but I got the following issue: https://github.com/caoxiemeihao/nuxt-plugin-electron/issues/1

**Comment by zavodnyrichard** [comment]
I also made it working, however the issue is the packaging for production.

**Comment by zavodnyrichard** [comment]
Okay so I used the Electron Builder and has been able to pack the application. What you have to do:

in main.js change the entry point to this:
`mainWindow.loadFile(path.join(__dirname, '.output', 'public', 'index.html'))`

then create .env file in root of the project folder and put this enviroment variable into it:
`NUXT_APP_BASE_URL: ./`

_*the default of that env var is "/"_

Then you also have to adjust your "scripts" object in package.lock: 
```
"scripts": {
    "build": "nuxt build",
    "dev": "nuxt dev",
    "generate": "nuxt generate",
    "preview": "nuxt preview",
    "postinstall": "nuxt prepare",
    "start": "concurrently \"nuxt dev\" \"electron .\"",
    "pack": "nuxt generate && electron-builder",
    "elbu": "electron-builder"
  },
```

and also add the "build" object with electron-builder configuration:
```
"build": {
    "appId": "my-great-electron-nuxt-app",
    "win": {
      "target": "portable"
    }
  }
```

and then just run this command in terminal:
`npm run pack`

This will pack generate Nuxt application into single entry point html file (".output/public/index.html") and also pack pack your application.

Also make sure that you use the <NuxtLink> component for any inter page navigation so the Vue Router behind is engaged, otherwise your pages won't be accessible and this will result in "blank screen".

I'm gonna make some "boilerplate" later today.

Hope this helps!

**Reply by gurvancampion** [reply]
Thanks for sharing, indeed it's working now, `NUXT_APP_BASE_URL="./"` did the trick.
However as you mentioned there is a problem with the navigation if you don't use NuxtLink (Vue router under the hood). Also when you refresh the App by pressing CMD+SHIFT+R (or CTRL+SHIFT+R on windows) it will display a blank screen. The path seems wrong: 
![CleanShot 2022-12-08 at 10 28 51](https://user-images.githubusercontent.com/32047986/206410961-a2988078-31b1-4d7a-8ce7-95ced96d57cf.png)

I made a minimal reproduction here: https://github.com/gurvan-guss/nuxt-electron-sample
I will try to see check what could be wrong here

**Reply by zavodnyrichard** [reply]
> Also when you refresh the App by pressing CMD+SHIFT+R (or CTRL+SHIFT+R on windows) it will display a blank screen.

I see, got the same issue. 

Also maybe you should disable asar (package.json):
```
"build": {
    ...
    "asar": false
  }
```

**Comment by zavodnyrichard** [comment]
It's weird. My application works great, however when I firstly start it, I get the 404 Page saying that it can't find the page and I have to click on the button "Go back home" to get to my index. So the Nuxt 3 and Router is working, however I gotta somehow bypass that initial 404. Wonder why this is happening... When I put my .output/public as entry point for my local web server (Laragon with Apache) it works like a charm. Also the reloading with CTRL+Shift+R, wonder if this might be problem with CORS? Not really sure, I'm tired of it already a bit.

**Comment by zavodnyrichard** [comment]
Just found a way to run Nitro on background of the packaged application:

It requires following changes:

package.json:
```
"scripts": {
    ...
    "server": "node .output/server/index.mjs",
    "pack": "nuxt build && electron-builder"
    ...
}
```

In main.js (electron main script) add these lines at the top:
```
const { exec } = require('node:child_process');
exec('npm run server');
```

and then change the entry point: 
`mainWindow.loadURL('http://localhost:3000')`

Now all you got to do is to to run `npm run pack'.

Works perfectly!

**Reply by ShuHado** [reply]
> Just found a way to run Nitro on background of the packaged application:
> 
> It requires following changes:
> 
> package.json:
> 
> ```
> "scripts": {
>     ...
>     "server": "node .output/server/index.mjs",
>     "pack": "nuxt build && electron-builder"
>     ...
> }
> ```
> 
> In main.js (electron main script) add these lines at the top:
> 
> ```
> const { exec } = require('node:child_process');
> exec('npm run server');
> ```
> 
> and then change the entry point: `mainWindow.loadURL('http://localhost:3000')`
> 
> Now all you got to do is to to run `npm run pack'.
> 
> Works perfectly!


Yes I seen this solution but it works only if you have Node.js installed on your computer... I search solution to run it without needing to install Node.js. And if anyone knows how to include prisma in the electron-builder build and have it operational in the created executable, I'm interested.

**Comment by gurvancampion** [comment]
I spend sometime again trying to build and pack nuxt in SPA mode with `ssr: false`. It works fine when you don't have a pages directory but when you specify one you get this initial `404 page not found`. However, after clicking on `Go Back Home` everything works fine. This is the last thing blocking me, I also cannot figure out what is causing this `404 page not found`.
Regarding the reloading in production with `CMD+SHIFT+R`, we could prevent it by removing it. In my case I added this to redirect automatically to the entry file `index.html` when it cannot load the correct path:

```ts
  win.webContents.on('did-fail-load', () => {
    if (app.isPackaged) win?.loadFile(path.join(distPath, 'index.html'))
  })
```

Here is a repository to reproduce the 404 error, just run `pnpm i` then `pnpm run build` and you will get `404 page not found`: https://github.com/gurvan-guss/nuxt-electron-sample

**Reply by zavodnyrichard** [reply]
Hello Gurvan, have you found any solution for this "404 page not found" error?

**Reply by gurvancampion** [reply]
Hi @zavodnyrichard no I didn't find any solution concerning the 404. I cannot figure out what is causing the issue

**Comment by gurvancampion** [comment]
I found the problem that caused the 404, here is the solution: https://github.com/nuxt/framework/issues/9663#issuecomment-1364018088
